### PR TITLE
[inductor] Decompose shard_dim_alltoall via regular decomposition

### PR DESCRIPTION
**Pull request by armijo1122**
Summary:
- add an opt-in regular Inductor decomposition for `_dtensor.shard_dim_alltoall` into layout ops plus `_c10d_functional.all_to_all_single`/`wait_tensor`
- decompose only when the dims are valid, the dtype is supported, the process group size is resolvable, and the local shard dim is evenly divisible by the process group size; otherwise keep the original op via `NotImplemented`
- register an autograd formula for `_dtensor::shard_dim_alltoall` that swaps `gather_dim`/`shard_dim` in backward, so compiled/AOTAutograd DTensor redistributes do not silently drop this gradient. This registration is always-on and is not gated by `TORCHINDUCTOR_DECOMPOSE_SHARD_DIM_ALLTOALL`.
- add unit coverage for exposing the c10d collective, local semantic equivalence, fallback cases, and a real `.backward()` through the registered autograd formula

This makes the all-to-all communication visible as standard functional collectives so later FX collective bucketing/overlap passes can reason about it. The decomposition is disabled by default and guarded by `TORCHINDUCTOR_DECOMPOSE_SHARD_DIM_ALLTOALL=1`.

Note: the autograd registration is separable from the opt-in decomposition, but is included here because regular decomposition is selected before AOTAutograd and `_dtensor::shard_dim_alltoall` needs a traceable backward formula when the op appears in compiled backward graphs.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

**Comment by armijo1122** [comment]
@eellison @IvanKobzarev Could you help review this PR when you have some time, thanks!

**Comment by jansel** [comment]
@claude review these changes

**Comment by armijo1122** [comment]
@jansel Thanks for the detailed review!

  I updated the pass to address the main correctness concerns:

  - The decomposition now only applies when `shard_dim_size % group_size == 0` can be guarded as true. Otherwise it keeps the original `_dtensor.shard_dim_alltoall` fallback, so we no longer try to invent uneven
  split semantics that may diverge from eager.
  - Replaced the inserted `aten.contiguous.default` nodes with `aten.clone.default(..., memory_format=torch.contiguous_format)`, which has an Inductor lowering.
  - Changed the inserted `_c10d_functional.all_to_all_single` split sizes from empty lists to explicit `[1] * group_size` for both input/output splits. After the pre-view/movedim, dim0 is exactly the process-
  group dimension, so each rank sends/receives one dim0 slice. This also makes fake/meta propagation produce the expected shape.
  - Added comments documenting the padded-even contract and that the graph is rank-specialized.
  - Added a lightweight numerical semantics test. It compares the decomposition against the local `_dtensor.shard_dim_alltoall` semantics expressed as `cat(inputs, gather_dim)` followed by `chunk(...,
  shard_dim)`, and covers a nontrivial 3D case with `gather_dim > shard_dim`.
  - Existing graph tests also check the generated collectives/wait nodes, split sizes, shape rewrites, skip-on-non-divisible behavior, and that `aten.contiguous.default` is not introduced.

  I did not add a full compile/runtime test in this PR because exercising this path end-to-end requires a real/fake distributed process group setup and tends to make the test significantly heavier/flakier for
  what is primarily a post-grad FX rewrite. The pass-level tests now cover the generated graph structure and the core tensor permutation semantics directly, while the actual lowering-sensitive issue was addressed
  by using `aten.clone.default(..., memory_format=torch.contiguous_format)`, which is already a supported Inductor lowering.

**Comment by armijo1122** [comment]
@aorenste @eellison @jansel Could you help review this PR again when you have some time, thanks!

**Comment by armijo1122** [comment]
@jansel These CI failures seem unrelated, could you help check again when you have some time, thanks!

**Comment by armijo1122** [comment]
@eellison @aorenste @jansel Could you help review this PR when you have a chance？I’d really appreciate your feedback.

**Comment by jansel** [comment]
@claude review these changes again

**Comment by armijo1122** [comment]
@jansel Thanks for your review! I just made some fix according to the advice from claude, please review again when you have some time

File: torch/_inductor/fx_passes/post_grad.py [hunk 1]
```diff
@@ -89,6 +93,124 @@
...
+                torch.ops.aten.clone.default,
+                args=(post_movedim,),
+                kwargs={"memory_format": torch.contiguous_format},
+            )
+            post_view = graph.call_function(
+                torch.ops.aten.view.default,
+                args=(post_clone, post_view_shape),
+            )
+
+        node.replace_all_uses_with(post_view)
```
Comment (jansel) [inline torch/_inductor/fx_passes/post_grad.py]: Can we do this with a regular decomp (in aot_autograd) rather than a pattern matcher pass?

**Review by jansel — CHANGES_REQUESTED** [review]
Can you explain more about why this is needed?  What is the use case?  Where/who will use this?  What other options were considered?  What does this enable or make better?

**Comment by armijo1122** [comment]
@jansel Thanks for the suggestion. I updated this to use a regular Inductor decomposition instead of a post-grad pattern matcher pass.

  The motivation here is to expose `_dtensor.shard_dim_alltoall` as explicit layout ops plus `_c10d_functional.all_to_all_single` / `wait_tensor`. This makes the communication visible to the existing Inductor
  collective handling, including later bucketing / overlap passes, instead of leaving it as an opaque DTensor op.

  One detail I found while moving this earlier is that `_dtensor.shard_dim_alltoall` did not have an autograd formula. In a training compile path, AOTAutograd can see the raw op before the decomposition is
  applied for the compiled forward/backward graphs, so the input gradient can be treated as `None`. To make the decomposition safe for training, I added the natural backward formula: the backward is another
  `shard_dim_alltoall` with `gather_dim` and `shard_dim` swapped.

  This revision:
  - removes the post-grad FX pattern rewrite
  - adds an opt-in regular decomposition for `_dtensor.shard_dim_alltoall`
  - keeps the decomposition conservative: enabled only by config, skips complex dtype, and only decomposes when the local shard dimension is evenly divisible by the group size
  - adds autograd registration for `_dtensor.shard_dim_alltoall`
  - updates tests to cover the decomposition structure, skip paths, local semantic equivalence, and the backward dim swap

  I also validated the compiled training path locally on 8 GPUs by comparing forward output and input gradients against an all-gather/chunk reference for several shapes; both forward and backward max error were
  0.

File: torch/_inductor/decomposition.py [hunk 2]
```diff
@@ -156,6 +160,72 @@ def register_decomposition(
...
+
+    try:
+        from torch.distributed.distributed_c10d import (
+            _get_group_size_by_name,
+            _resolve_process_group,
+            GroupName,
+        )
+
+        group = _resolve_process_group(GroupName(group_name))
+        group_size = _get_group_size_by_name(group)
```
Comment (aorenste) [inline torch/_inductor/decomposition.py]: [cleanup] Redundant resolve: `_get_group_size_by_name` already accepts a `GroupName` and resolves the group internally (`distributed_c10d.py:1197-1201`), and `group` isn't used for anything else here. Collapse to:

```python
        group_size = _get_group_size_by_name(GroupName(group_name))
```

File: test/inductor/test_decompose_shard_dim_alltoall.py [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+        self.assertIs(result[0], expected)
+        self.assertEqual(result[1:], (None, None, None))
+        shard_dim_alltoall.assert_called_once()
+        args = shard_dim_alltoall.call_args.args
+        self.assertIs(args[0], grad)
+        self.assertEqual(args[1:], (0, 2, "test_pg"))
+
+
+if __name__ == "__main__":
+    if IS_LINUX:
```
Comment (aorenste) [inline test/inductor/test_decompose_shard_dim_alltoall.py]: [cleanup] Guarding `run_tests()` on `IS_LINUX` makes this file a silent no-op (exit 0, nothing run) rather than a visible skip when it is executed as `__main__` on a non-Linux runner -- so the inductor Mac/Windows jobs would report pass without actually running these tests. This suite is pure fake-tensor + mocks with no process group, so there is no Linux dependency; drop the guard:

```python
if __name__ == "__main__":
    run_tests()
```

**Review by aorenste — CHANGES_REQUESTED** [review]
(Reviewed by me, assisted by AI)

[blocker] The description no longer matches the diff, and closing that gap is exactly what jansel's open CHANGES_REQUESTED is asking for:

- It describes "an opt-in Inductor post-grad pass," but the current implementation is a regular `@register_decomposition` in `torch/_inductor/decomposition.py` (there is no post-grad pass), and it doesn't mention the `torch.library.register_autograd("_dtensor::shard_dim_alltoall", ...)` added in `torch/distributed/tensor/_collective_utils.py`.
- That autograd registration is not gated by `decompose_shard_dim_alltoall`; it ships always-on. It therefore changes the op's differentiability in the compiled/AOTAutograd path for every DTensor Shard->Shard redistribute, independent of the opt-in flag. (Eager is unaffected: redistribute runs the op inside `Redistribute.forward`, an `autograd.Function` whose forward disables grad.) Per the author's own note this fixes gradients that were previously silently dropped under compile, so it is likely correct and desirable -- but as a separable, BC-relevant change it should (a) be called out in the description (that directly answers jansel's why/scope question), (b) be considered as its own PR independent of the experimental decomposition, and (c) have a test that actually drives `.backward()` through the registered op. Neither backward-named test does that today: `test_backward_formula_matches_original_api_semantics_locally` re-runs the forward decomposition with `gather_dim`/`shard_dim` swapped (so the name is misleading), and `test_shard_dim_alltoall_backward_swaps_dims` calls the helper directly with a mocked op -- so a wrong `setup_context` or return arity would still pass CI.

**Comment by armijo1122** [comment]
@aorenste Thanks for the careful review!

  I updated the PR to address the comments:

  1. Removed the redundant process-group resolve in the decomposition. The code now directly calls `_get_group_size_by_name(GroupName(group_name))`, since `_get_group_size_by_name` already
  handles resolving the group internally.

  2. Dropped the `IS_LINUX` guard around `run_tests()`. The test file only uses fake tensor / mocks / a local CPU impl and does not require Linux-specific distributed setup, so it should run
  normally instead of becoming a silent no-op on non-Linux runners.

  3. Reworked the autograd test to exercise the actual registered autograd path. The new test calls `torch.ops._dtensor.shard_dim_alltoall.default(...).sum().backward()` with a scoped CPU
  impl, and verifies that backward invokes `shard_dim_alltoall` with `gather_dim` / `shard_dim` swapped and produces the expected input gradient.

  4. Updated the PR title and description to match the current implementation: this is now a regular Inductor decomposition, not a post-grad pass. I also called out that the
  `_dtensor::shard_dim_alltoall` autograd registration is always-on and separate from the opt-in decomposition flag.

  Regarding the autograd registration: I kept it always registered rather than gated by the Inductor config, since it represents the operator’s differentiability semantics, while the config
  only controls whether Inductor decomposes the op for communication visibility.

File: test/inductor/test_decompose_shard_dim_alltoall.py [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+        inp = torch.empty(shape, dtype=dtype)
+        decomp = self._decomp()
+
+        def fn(x):
+            return decomp(x, gather_dim, shard_dim, "test_pg")
+
+        resolve_group, get_group_size = self._group_mocks(group_size)
+        with config.patch({"decompose_shard_dim_alltoall": True}):
+            with resolve_group, get_group_size:
+                return make_fx(fn, tracing_mode="fake")(inp)
```
Comment (aorenste) [inline test/inductor/test_decompose_shard_dim_alltoall.py]: [cleanup] `_trace_decomp` traces the decomposition *function* directly (`fn` calls `decomp(x, ...)`), so `test_decomposition_exposes_c10d_collective`'s `assertNotIn(shard_dim_alltoall.default)` is trivially true -- the op is never placed in the graph to begin with. Nothing in the suite drives the op *through* the decomposition table, which is the exact routing this PR relies on at compile time (only the manual 8-GPU run exercises it). Trace the op instead:

    def fn(x):
        return torch.ops._dtensor.shard_dim_alltoall.default(
            x, gather_dim, shard_dim, "test_pg"
        )
    ...
    return make_fx(
        fn, decomposition_table=select_decomp_table(), tracing_mode="fake"
    )(inp)

Then `assertNotIn` is meaningful (op present pre-decomp, gone post) and the table wiring is locked in.

File: test/inductor/test_decompose_shard_dim_alltoall.py [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+
+        return outputs
+
+    def _decomp(self):
+        return select_decomp_table()[torch.ops._dtensor.shard_dim_alltoall.default]
+
+    def _group_mocks(self, group_size: int):
+        return (
+            mock.patch(
+                "torch.distributed.distributed_c10d._resolve_process_group",
```
Comment (aorenste) [inline test/inductor/test_decompose_shard_dim_alltoall.py]: [cleanup] (minor) This `_resolve_process_group` patch is dead: the decomposition resolves group size via `_get_group_size_by_name(GroupName(group_name))` (patched just below) and never calls `_resolve_process_group`. It is a leftover from the removed post-grad pass. Dropping it means returning only the `_get_group_size_by_name` patch and updating the three `with resolve_group, get_group_size:` call sites to `with get_group_size:`.

File: torch/_inductor/decomposition.py [hunk 6]
```diff
@@ -156,6 +160,70 @@ def register_decomposition(
...
+    ndim = inp.dim()
+    gather_dim = gather_dim + ndim if gather_dim < 0 else gather_dim
+    shard_dim = shard_dim + ndim if shard_dim < 0 else shard_dim
+    if not (0 <= gather_dim < ndim and 0 <= shard_dim < ndim):
+        return NotImplemented
+    if gather_dim == shard_dim:
+        return NotImplemented
+
+    try:
+        from torch.distributed.distributed_c10d import (
```
Comment (aorenste) [inline torch/_inductor/decomposition.py]: [future] It's possible that distributed wasn't compiled in... but then I guess you just wouldn't set config.decompose_shard_dim_alltoall - but if we ever make it default to True then we might want to cache the import...

**Review by aorenste — APPROVED** [review]
(Reviewed by me, assisted by AI)

File: test/inductor/test_decompose_shard_dim_alltoall.py (hunk 4 above)
Comment (armijo1122) [reply test/inductor/test_decompose_shard_dim_alltoall.py]: Thanks for your advice, cleanup is done

File: test/inductor/test_decompose_shard_dim_alltoall.py (hunk 5 above)
Comment (armijo1122) [reply test/inductor/test_decompose_shard_dim_alltoall.py]: Done, it's dead code

File: torch/_inductor/decomposition.py (hunk 6 above)
Comment (armijo1122) [reply torch/_inductor/decomposition.py]: Thanks, that makes sense.

  My assumption here was that this path is only reachable when `config.decompose_shard_dim_alltoall` is explicitly enabled, and enabling this pass should imply that the workload is using DTensor / distributed
  collectives, so `torch.distributed` would already be available in that environment.

  Do you think it is reasonable to treat `decompose_shard_dim_alltoall=True` as requiring a distributed-enabled build, and keep the direct import here?

I think keeping the direct import here is reasonable for now, mainly to keep the implementation simple.

  Python module imports are already cached in `sys.modules`, so this should not repeatedly reload `torch.distributed.distributed_c10d`. This decomposition path is also not a hot runtime path: it only runs
  during tracing/compilation, and only when `config.decompose_shard_dim_alltoall` is explicitly enabled.

**Comment by armijo1122** [comment]
@aorenste @jansel These CI failures seem unrelated to current changes, could you help check again, thanks.

**Comment by jansel** [comment]
@claude review these changes again

**Review by jansel — APPROVED** [review]

**Comment by armijo1122** [comment]
@jansel Could you help skip these unrelated CI failures, thanks

**Comment by armijo1122** [comment]
@jansel @aorenste The failure looks related: `torch.ops._dtensor.shard_dim_alltoall` may be unavailable on some builds such as mac-arm64, so the unconditional decomposition registration fails at import time before the
  config check can run. Would it be acceptable to guard the registration with `getattr(torch.ops._dtensor, "shard_dim_alltoall", None)` and only register the decomposition when the op exists?

File: torch/_inductor/decomposition.py [hunk 7]
```diff
@@ -156,6 +160,72 @@ def register_decomposition(
     return decomp.register_decomposition(ops, decompositions)
 
 
+if torch.distributed.is_available():
```
Comment (armijo1122) [inline torch/_inductor/decomposition.py:163]: I guarded `shard_dim_alltoall_decomp` registration with `torch.distributed.is_available()`, since `_dtensor::shard_dim_alltoall` is registered from distributed/c10d and may be unavailable in builds without
  distributed support, e.g. default macOS builds. Hopefully this addresses the import-time failures in those environments.

File: test/inductor/test_decompose_shard_dim_alltoall.py [hunk 8]
```diff
@@ -0,0 +1,283 @@
...
+import torch
+from torch import fx
+from torch._dynamo.utils import counters
+from torch._inductor import config
+from torch._inductor.decomposition import select_decomp_table
+from torch._inductor.test_case import run_tests, TestCase
+from torch.fx.experimental.proxy_tensor import make_fx
+
+
+@unittest.skipIf(not torch.distributed.is_available(), "requires distributed support")
```
Comment (armijo1122) [inline test/inductor/test_decompose_shard_dim_alltoall.py:16]: @jansel @aorenste The test
  should also avoid accessing distributed-only APIs in builds where distributed is not compiled/enabled. This was the only CI failure I found related to this PR; the other current failures look unrelated to this change. Could you help trigger CI run and merge again, thanks.
